ledger/apply/asset.go

```diff
@@ -308,39 +337,51 @@ func AssetTransfer(ct transactions.AssetTransferTxnFields, header transactions.H
 			return fmt.Errorf("cannot close asset by clawback")
 		}
 
-		// Fetch the sender balance record. We will use this to ensure
-		// that the sender is not the creator of the asset, and to
-		// figure out how much of the asset to move.
-		snd, err := balances.Get(source, false)
+		record, err := balances.Get(source, false)
 		if err != nil {
 			return err
 		}
 
+		if record.TotalAssets == 0 {
+			return fmt.Errorf("cannot close asset holding on account %s : account is not opted in asset %d ", source.String(), ct.XferAsset)
+		}
+
+		// Fetch the sender asset data. We will use this to ensure
+		// that the sender is not the creator of the asset, and to
+		// figure out how much of the asset to move.
+
 		// The creator of the asset cannot close their holding of the
 		// asset. Check if we are the creator by seeing if there is an
 		// AssetParams entry for the asset index.
-		if _, ok := snd.AssetParams[ct.XferAsset]; ok {
+		ok, err := balances.HasAssetParams(source, ct.XferAsset)
+		if err != nil {
+			return err
+		}
+		if ok {
 			return fmt.Errorf("cannot close asset ID in allocating account")
 		}
 
 		// Fetch our asset holding, which should exist since we're
 		// closing it out
-		sndHolding, ok := snd.Assets[ct.XferAsset]
+		sndHolding, ok, err := balances.GetAssetHolding(source, ct.XferAsset)
+		if err != nil {
+			return err
+		}
```

Comment:
It seems like line 356 _ought_ to make line _366_ fast. But I don't think it does. From what I see, looking up a params does not put the holding info into a cache. Am I missing it?

Reply:
hmm. let's trace it:
the call to `balances.HasAssetParams` calls `roundCowState.HasAssetParams`, which calls `roundCowState.lookupAssetParams`, which calls `roundCowBase.lookupAssetParams` which calls `roundCowBase.updateAssetResourceCache`.

did I miss something ?

Reply:
Yes, I think I figured it out eventually.  I was expecting a cache at each cow level, but it's only in the base.  And that seems like a good idea now that I understand it.